components/eventing-controller/controllers/subscription/nats/reconciler.go

```diff
@@ -173,139 +172,201 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
-		desiredSubscription.ObjectMeta.Finalizers = append(desiredSubscription.ObjectMeta.Finalizers, Finalizer)
-		if err := r.Update(context.Background(), desiredSubscription); err != nil {
+	if !utils.ContainsString(subscription.ObjectMeta.Finalizers, Finalizer) {
+		subscription.ObjectMeta.Finalizers = append(subscription.ObjectMeta.Finalizers, Finalizer)
+		if err := r.Update(context.Background(), subscription); err != nil {
 			log.Errorw("add finalizer to subscription failed", "error", err)
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
 			return ctrl.Result{}, err
 		}
 		log.Debug("add finalizer to subscription succeeded")
-		result.Requeue = true
+		return ctrl.Result{Requeue: true}, nil
 	}
 
-	if result.Requeue {
-		return result, nil
+	// Check for valid sink
+	if err := r.sinkValidator(ctx, r, subscription); err != nil {
+		log.Errorw("sink URL validation failed", "error", err)
+		if err := r.syncSubscriptionStatus(ctx, subscription, false, false, err.Error()); err != nil {
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		// No point in reconciling as the sink is invalid
+		return ctrl.Result{}, nil
 	}
 
-	_, err = r.Backend.SyncSubscription(desiredSubscription, r.eventTypeCleaner)
+	// Synchronize Kyma subscription to NATS backend
+	subscriptionStatusChanged, err := r.Backend.SyncSubscription(subscription, r.eventTypeCleaner)
 	if err != nil {
 		log.Errorw("sync subscription failed", "error", err)
-		if err := r.syncSubscriptionStatus(ctx, desiredSubscription, false, err.Error()); err != nil {
+		if err := r.syncSubscriptionStatus(ctx, subscription, false, false, err.Error()); err != nil {
+			if k8serrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
 			return ctrl.Result{}, err
 		}
 		return ctrl.Result{}, err
 	}
 	log.Debug("create NATS subscriptions succeeded")
 
 	// Update status
-	if err := r.syncSubscriptionStatus(ctx, desiredSubscription, true, ""); err != nil {
+	if err := r.syncSubscriptionStatus(ctx, subscription, true, subscriptionStatusChanged, ""); err != nil {
+		if k8serrors.IsConflict(err) {
+			// Requeue the Pod to try to reconciliate again,
+			return ctrl.Result{Requeue: true}, nil
+		}
 		return ctrl.Result{}, err
 	}
 
-	return result, nil
+	return ctrl.Result{}, nil
 }
 
 // syncSubscriptionStatus syncs Subscription status
-// subsConfig is the subscription configuration that was applied to the subscription. It is set only if the
-// isNatsSubReady is true.
-func (r *Reconciler) syncSubscriptionStatus(ctx context.Context, sub *eventingv1alpha1.Subscription, isNatsSubReady bool, message string) error {
-	desiredSubscription := sub.DeepCopy()
+func (r *Reconciler) syncSubscriptionStatus(ctx context.Context, sub *eventingv1alpha1.Subscription, isNatsSubReady bool, forceUpdateStatus bool, message string) error {
```

Comment:
This method is really **hard to read** and **understand**. As far as I see there is only **one condition** which is used for NATS: `eventingv1alpha1.ConditionSubscriptionActive`

Can't we simplify the method ?
- there is no need for a loop, there is only one condition
- we could create the `desired` conditions and then `compare` them to the `current` conditions (of course without comparing `LastTransitinTime`) and update if both are not equal.

Reply:
As discussed, the original design was to be ready to have multiple conditions which should also "explain" why the global "ready" status is false (like in the BEB case). It could happen that we will add more conditions for JS backend.
Agree, we should redesign this piece of code..